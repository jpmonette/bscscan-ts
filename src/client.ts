import { Accounts } from "./modules/accounts";
import { Contracts } from "./modules/contracts";
import { Stats } from "./modules/stats";
import { Transactions } from "./modules/transactions";
import { APIResponse, BscScanOptions } from "./typings";

/**
 * The default production API base url
 */
const BASE_URL = "https://api.bscscan.com/api?";

/**
 * A BscScan API client
 *
 * @class BscScan
 */
class BscScan {
  apikey: string | undefined;
  baseUrl: string;

  accounts: Accounts;
  contracts: Contracts;
  transactions: Transactions;
  stats: Stats;

  /**
   * Creates an instance of BscScan.
   * @param {BscScanOptions} opts
   * @memberof BscScan
   */
  constructor(opts: BscScanOptions) {
    const { apikey } = opts;

    this.apikey = apikey;
    this.baseUrl = opts.baseUrl || BASE_URL;

    this.accounts = new Accounts(this);
    this.contracts = new Contracts(this);
    this.stats = new Stats(this);
    this.transactions = new Transactions(this);
  }

  /**
   * Creates a new HTTP request
   *
   * @param {string} [method="GET"]
   * @param {string} module
   * @param {string} action
   * @param {Record<string, any>} [opts={}]
   * @returns {Request}
   * @memberof BscScan
   */
  newRequest(method = "GET", module: string, action: string, opts: Record<string, any> = {}): Request {
    const params = new URLSearchParams({ module, action });

    if (this.apikey) {
      params.set("apikey", this.apikey);
    }

    for (const [key, value] of Object.entries(opts)) {
      if (Array.isArray(value)) {
        params.set(key, value.join(","));
      } else {
        params.set(key, value);
      }
    }

    const headers: Headers = new Headers();
    headers.set("Content-Type", "application/json; charset=UTF-8");

    const init: RequestInit = { method, headers };

    const url = new URL("/api?" + params.toString(), baseUrl);

    const request = new Request(url.toString(), init);
    return request;
  }

  /**
   * Executes a simplified API query
   *
   * @param {string} module
   * @param {string} action
   * @param {Record<string, any>} [opts={}]
   * @returns {Promise<any>}
   * @memberof BscScan
   */
  async query(module: string, action: string, opts: Record<string, any> = {}): Promise<any> {
    return this.do(this.newRequest("GET", module, action, opts));
  }

  /**
   * Executes an HTTP request
   *
   * @template T
   * @param {Request} request
   * @returns {Promise<T>}
   * @memberof BscScan
   */
  async do<T>(request: Request): Promise<T> {
    const response: Response = await fetch(request);
    const responseBody: APIResponse<T> = await response.json();

    if (responseBody.status === "0") {
      throw new Error(responseBody.message + " - " + responseBody.result);
    }

    return responseBody.result;
  }
}

export { BscScan };
