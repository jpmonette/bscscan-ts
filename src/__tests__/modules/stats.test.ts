import "isomorphic-fetch";
import { BscScan } from "../../client";
import { server } from "../server";
import { API_KEY } from "../setup";

describe("stats", () => {
  const client = new BscScan({ apikey: API_KEY });

  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it("should getBNBLastPrice", async () => {
    expect(await client.stats.getBNBLastPrice()).toMatchSnapshot();
  });

  it("should getBNBTotalSupply", async () => {
    expect(await client.stats.getBNBTotalSupply()).toMatchSnapshot();
  });

  it("should getValidators", async () => {
    expect(await client.stats.getValidators()).toMatchSnapshot();
  });
});
