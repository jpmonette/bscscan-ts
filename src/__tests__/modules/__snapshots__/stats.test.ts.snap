// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats should getBNBLastPrice 1`] = `
Object {
  "ethbtc": "0.009928",
  "ethbtc_timestamp": "1639590487",
  "ethusd": "582.6",
  "ethusd_timestamp": "1639590487",
}
`;

exports[`stats should getBNBTotalSupply 1`] = `"22479013170398300000000000"`;

exports[`stats should getValidators 1`] = `
Array [
  Object {
    "validatorAddress": "0x9f8ccdafcc39f3c7d6ebf637c9151673cbc36b88",
    "validatorName": "",
    "validatorStatus": "0",
    "validatorVotingPower": "43379676392570",
    "validatorVotingPowerProportion": "0.0617",
  },
  Object {
    "validatorAddress": "0x2465176c461afb316ebc773c61faee85a6515daa",
    "validatorName": "",
    "validatorStatus": "0",
    "validatorVotingPower": "38039845465042",
    "validatorVotingPowerProportion": "0.0541",
  },
]
`;
